### ISSUE TITLE
Rotary Embeddings not properly initialized when loading model with from_pretrained()

### ISSUE THREAD

**Issue by fire537** [macwiatrak/Bacformer#33]
Problem:
  When loading Bacformer models (both Base and Large) using AutoModel.from_pretrained(), the Rotary Embeddings buffers are not properly initialized, causing the model to output NaN
  values during inference.

  Root Cause:
  The rotary embeddings are registered with persistent=False:

  Base model (modeling_base.py line 470-471):
  self.register_buffer("freqs_cos", freqs_cos, persistent=False)
  self.register_buffer("freqs_sin", freqs_sin, persistent=False)

  Large model (modeling_large.py line 358):
  self.register_buffer("inv_freq", inv_freq, persistent=False)

  persistent=False means these buffers are NOT saved to the checkpoint. When loading with from_pretrained(), HuggingFace's loading mechanism does not automatically re-initialize these
   values, resulting in NaN or garbage values.

  This leads to incorrect attention scores (inf/-inf) and NaN outputs.

  Actual values after loading:
  - freqs_cos: Expected [-1.0, 1.0], Actual: NaN or extreme values
  - freqs_sin: Expected [-1.0, 1.0], Actual: 0 or extreme values
  - inv_freq: Expected [~0.0001, 1.0], Actual: 0 or extreme values

  Workaround:

  For Base model:
  from bacformer.modeling.modeling_base import precompute_freqs_cis

  config = bacformer_model.config
  freqs_cos, freqs_sin = precompute_freqs_cis(
      config.hidden_size // config.num_attention_heads,
      int(config.max_position_embeddings * 1.5)
  )
  bacformer_model.encoder.register_buffer("freqs_cos", freqs_cos.to(device), persistent=False)
  bacformer_model.encoder.register_buffer("freqs_sin", freqs_sin.to(device), persistent=False)

  For Large model:
  config = bacformer_model.config
  dim = config.hidden_size // config.num_attention_heads
  inv_freq = 1.0 / (10000.0 ** (torch.arange(0, dim, 2, device=device, dtype=torch.float32) / dim))

  for layer in bacformer_model.encoder.layers:
      layer.attn.rotary.register_buffer("inv_freq", inv_freq.clone(), persistent=False)

**Comment by choijein0410** [comment]
## Independent reproduction: version boundary, silent corruption, and current Large revision

Thank you @fire537 for the earlier public root-cause report. I independently
reproduced #33 and ran additional controls to distinguish checkpoint-weight
corruption from missing deterministic RoPE buffers. The results below confirm
the report and add a silent finite-output case, pinned checkpoint revisions, a
version matrix, and analytic-repair controls.

### Summary and maintainer request

- **Base and Large can return finite but wrong outputs under tested
  Transformers 5 loaders.** Output-finiteness checks do not catch the defect.
- In clean same-PyTorch controls, analytically reconstructing only the missing
  rotary buffers restores bit-identical agreement with Transformers 4.56.2.
- The current Large revision remains affected after the separate `8b84601`
  embedding LayerNorm change.
- A temporary tested workaround is `transformers==4.56.2`.

Would the maintainers accept a PR that reconstructs the buffers after load,
asserts their analytic values, and temporarily adds `transformers<5` (or an
equivalent model-card warning) until compatible releases are verified?

### Environment and immutable revisions

- macOS arm64, CPU
- Base decisive control: Python 3.11.15, PyTorch 2.5.1,
  Transformers 5.14.1, Accelerate 1.10.1, FP32
- Large cross-version control: Python 3.11.15, PyTorch 2.5.1,
  Transformers 4.56.2 versus 5.4.0
- Base MAG:
  `macwiatrak/bacformer-masked-MAG@17b98674d11b491aa98ce7a3809061523b9b3752`
- Base complete genomes:
  `macwiatrak/bacformer-masked-complete-genomes@6c8d83deabe7e845627a67213aef8b726ce9c9c9`
- Large MAG:
  `macwiatrak/bacformer-large-masked-MAG@589fda44a5ef9f75ccc1cf7e1a6283e1075db34d`

The Large revision is current after the separate LayerNorm change
(`8b84601` / `589fda44`). That code change only reordered the embedding
Linear/LayerNorm operations; it retained `inv_freq` with `persistent=False`
and no rotary reset, while the checkpoint LFS weight object stayed unchanged.
The current revision is still affected by this RoPE loading issue.

### Controlled findings

1. **Checkpoint contents:** neither Base checkpoint contains
   `encoder.freqs_cos` or `encoder.freqs_sin`; the Large checkpoint contains no
   `inv_freq` keys. This is expected from `persistent=False`, but requires
   correct reconstruction during loading.
2. **Version dependence:** Base buffers were analytically exact under
   Transformers 4.38.2, 4.44.2, and 4.56.2. The tested Transformers 5 loads
   produced invalid buffers. Transformers 5.0.0rc3 and 5.1.0 also showed a
   separate parameter-loading problem in this stack, so I do not use those
   releases for the buffer-only causal claim.
3. **Clean Base isolation:** with Transformers 5.14.1/FP32, all 162 backbone
   parameters matched the checkpoint exactly and shared the same complete
   parameter hash as the correct 4.56.2 load. Both RoPE tables were instead
   all zero. The output was fully finite but differed from 4.56.2 by
   `max_abs=0.454705`, `mean_abs=0.083556`, `RMSE=0.105263`.
4. **Fresh-process replication:** two separately launched 30-load cohorts on
   the same host/runtime (Transformers 5.4.0, Python 3.13.12, PyTorch 2.11.0)
   produced 23/60 finite-but-wrong Base outputs. In one cohort, 10/30 loads had
   both full RoPE tables exactly zero; all ten produced finite wrong outputs.
   The frequency is allocator/environment-specific and should not be
   generalized.
5. **Analytic repair:** all 90 modern repeated loads became finite after
   recomputation. Repaired Transformers 5.14.1 Base output was bit-identical
   to the known-good 4.x output with PyTorch held at 2.5.1.
6. **Scope:** the Base complete-genomes checkpoint is affected. For the
   current Large MAG revision under Transformers 5.4.0, all 30
   `rotary.inv_freq` buffers were exactly zero. The forward output was finite
   but differed from the same-PyTorch 4.56.2 reference by
   `max_abs=3.243294`, `mean_abs=0.681549`; repair made the two outputs
   bit-identical.
7. **Loader switch:** `low_cpu_mem_usage=False, device_map=None` did not fix
   Base or Large. Current Transformers 5 removes `low_cpu_mem_usage` from the
   kwargs before construction.

This supports a compatibility failure in which Transformers 5 materializes
non-persistent constructor-created meta buffers with empty storage and no
later Bacformer step reconstructs them. Depending on the allocation, the
failure can be loud (NaN) or silent (finite but wrong).

<details>
<summary>29-line Base reproducer and observed output</summary>

```python
import torch
import transformers
from transformers import AutoModel

print("torch=", torch.__version__, "transformers=", transformers.__version__)
model_id = "macwiatrak/bacformer-masked-MAG"
model = AutoModel.from_pretrained(
    model_id, revision="17b98674d11b491aa98ce7a3809061523b9b3752",
    trust_remote_code=True, dtype=torch.float32
).eval()
dim = model.config.hidden_size // model.config.num_attention_heads
end = int(model.config.max_position_embeddings * 1.5)
freq = 1 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
phase = torch.outer(torch.arange(end).float(), freq)
ref_cos, ref_sin = torch.cos(phase), torch.sin(phase)
print("transformers buffer check:")
for name, ref in [("freqs_cos", ref_cos), ("freqs_sin", ref_sin)]:
    buf = getattr(model.encoder, name).float().cpu()
    print(name, "all-zero=", torch.count_nonzero(buf).item() == 0,
          "NaNs=", torch.isnan(buf).sum().item(),
          "max|Δ|=", (buf - ref).abs().nan_to_num(posinf=torch.inf).max().item())
length = 11
x = torch.arange(length * model.config.hidden_size).reshape(1, length, -1).float() % 97 / 97
special = torch.full((1, length), model.config.prot_emb_token_id)
with torch.inference_mode():
    y = model(protein_embeddings=x, special_tokens_mask=special,
              token_type_ids=torch.zeros_like(special),
              attention_mask=torch.ones_like(special))[0]
print("output finite=", torch.isfinite(y).all().item())
```

On one clean Transformers 5.14.1/PyTorch 2.5.1 run this exits normally with:

```text
freqs_cos all-zero=True NaNs=0 max|Δ|=1.0
freqs_sin all-zero=True NaNs=0 max|Δ|=1.0
output finite=True
```

</details>

### Workaround and proposed invariant

- Temporary tested workaround: pin `transformers==4.56.2`.
- Alternatively, analytically reconstruct the Base tables or every Large
  `inv_freq` after loading and verify exact equality before inference.
- A durable fix could persist these small deterministic buffers or reconstruct
  them after materialization. A regression test should assert analytic buffer
  equality and 4.x/5.x output agreement; output finiteness alone is
  insufficient.

### Historical-paper boundary

I found no evidence that this specific Transformers-5 mechanism affected the
July 2025 paper. Transformers 5 postdates the preprint. I also pinned the four
8-July-2025 masked/causal × MAG/complete checkpoint revisions and loaded each
under Transformers 4.50.3/PyTorch 2.5.1: all RoPE buffers were analytically
exact, all forward outputs were finite, analytic repair was a bitwise no-op,
and checkpoint parameters matched 162/162 (masked) or 168/168 (causal). This is
not a full paper reproduction or a reconstruction of the authors' complete
runtime, but it strongly excludes this v5 loader path for the original paper.

Could the maintainers instead confirm whether any Large/BacBench embeddings or
evaluation artifacts generated after stable Transformers 5 became available
used a 5.x environment? Package versions, immutable model revisions, creation
dates, and artifact hashes would identify anything requiring regeneration. An
archived 2025 paper environment would still be useful for the reproducibility
record, but the present evidence points away from paper impact.

I can provide the complete version matrix, raw per-load JSON/CSV, and sanitized
logs if useful. If the proposed fix direction is acceptable, would a PR be
welcome?
